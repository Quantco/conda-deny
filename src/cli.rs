use clap::ArgAction;
use clap_verbosity_flag::{ErrorLevel, Verbosity};

use clap::Parser;

#[derive(Parser, Debug)]
#[command(name = "conda-deny", about = "Check and list licenses of pixi and conda environments", version = env!("CARGO_PKG_VERSION"))]
pub struct Cli {
    #[command(flatten)]
    pub verbose: Verbosity<ErrorLevel>,

    #[command(subcommand)]
    pub command: CondaDenyCliConfig,

    /// Path to the conda-deny config file
    #[arg(short, long, global = true)]
    pub config: Option<String>,
}

#[derive(clap::Subcommand, Debug)]
pub enum CondaDenyCliConfig {
    Check {
        /// Path to the pixi lockfile(s)
        #[arg(short, long)]
        lockfile: Option<Vec<String>>,

        /// Path to the conda prefix(es)
        #[arg(long, global = true)]
        prefix: Option<Vec<String>>,

        /// Platform(s) to check
        #[arg(short, long)]
        platform: Option<Vec<String>>,

        /// Environment(s) to check
        #[arg(short, long)]
        environment: Option<Vec<String>>,

        #[arg(short, long, action = ArgAction::SetTrue)]
        include_safe: bool,

        #[arg(short, long)]
        osi: Option<bool>,

        /// Ignore when encountering pypi packages instead of failing.
        #[arg(long)]
        ignore_pypi: Option<bool>,
    },
    List {
        /// Path to the pixi lockfile(s)
        #[arg(short, long)]
        lockfile: Option<Vec<String>>,

        /// Path to the conda prefix(es)
        #[arg(long, global = true)]
        prefix: Option<Vec<String>>,

        /// Platform(s) to list
        #[arg(short, long)]
        platform: Option<Vec<String>>,

        /// Environment(s) to list
        #[arg(short, long)]
        environment: Option<Vec<String>>,

        /// Ignore when encountering pypi packages instead of failing.
        #[arg(long)]
        ignore_pypi: Option<bool>,
    },
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_cli_check_include_safe() {
        let cli = Cli::try_parse_from(vec!["conda-deny", "check", "--include-safe"]).unwrap();
        match cli.command {
            CondaDenyCliConfig::Check { include_safe, .. } => {
                assert!(include_safe);
            }
            _ => panic!("Expected check subcommand with --include-safe"),
        }
    }

    #[test]
    fn test_cli_with_config() {
        let cli =
            Cli::try_parse_from(vec!["conda-deny", "list", "--config", "custom.toml"]).unwrap();
        assert_eq!(cli.config.as_deref(), Some("custom.toml"));
    }

    #[test]
    fn test_cli_with_config_new_order() {
        let cli =
            Cli::try_parse_from(vec!["conda-deny", "check", "--config", "custom.toml"]).unwrap();
        assert_eq!(cli.config.as_deref(), Some("custom.toml"));
        match cli.command {
            CondaDenyCliConfig::Check { include_safe, .. } => {
                assert!(!include_safe);
            }
            _ => panic!("Expected check subcommand with --config"),
        }
    }

    #[test]
    fn test_cli_with_check_arguments() {
        let cli = Cli::try_parse_from(vec!["conda-deny", "check", "--include-safe"]).unwrap();
        match cli.command {
            CondaDenyCliConfig::Check { include_safe, .. } => {
                assert!(include_safe);
            }
            _ => panic!("Expected check subcommand with --include-safe"),
        }
    }
}
