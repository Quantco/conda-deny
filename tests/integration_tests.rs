use assert_cmd::prelude::*;
use conda_deny::bundle::bundle;
use conda_deny::cli::CondaDenyCliConfig;
use conda_deny::{
    check::check, get_config_options, list::list, CondaDenyCheckConfig, CondaDenyConfig,
    CondaDenyListConfig,
};
use conda_deny::{CondaDenyBundleConfig, OutputFormat};
use rattler_conda_types::Platform;
use rstest::{fixture, rstest};
use std::io::Write;
use std::path::{Path, PathBuf};
use std::process::Command;
use tempfile::NamedTempFile;
use walkdir::WalkDir;

#[fixture]
fn list_config(
    #[default(None)] config: Option<PathBuf>,
    #[default(None)] lockfile: Option<Vec<PathBuf>>,
    #[default(None)] prefix: Option<Vec<PathBuf>>,
    #[default(None)] platform: Option<Vec<Platform>>,
    #[default(None)] environment: Option<Vec<String>>,
    #[default(None)] ignore_pypi: Option<bool>,
    #[default(Some(OutputFormat::Default))] output: Option<OutputFormat>,
) -> CondaDenyListConfig {
    let cli = CondaDenyCliConfig::List {
        lockfile,
        prefix,
        platform,
        environment,
        ignore_pypi,
        output,
    };

    let config = get_config_options(config, cli).unwrap();

    match config {
        CondaDenyConfig::List(list_config) => list_config,
        _ => panic!(),
    }
}

#[fixture]
fn bundle_config(
    #[default(None)] config: Option<PathBuf>,
    #[default(None)] lockfile: Option<Vec<PathBuf>>,
    #[default(None)] prefix: Option<Vec<PathBuf>>,
    #[default(None)] platform: Option<Vec<Platform>>,
    #[default(None)] environment: Option<Vec<String>>,
    #[default(None)] ignore_pypi: Option<bool>,
    #[default(None)] directory: Option<PathBuf>,
) -> CondaDenyBundleConfig {
    let cli = CondaDenyCliConfig::Bundle {
        lockfile,
        prefix,
        platform,
        environment,
        ignore_pypi,
        directory,
    };

    let config = get_config_options(config, cli).unwrap();

    match config {
        CondaDenyConfig::Bundle(bundle_config) => bundle_config,
        _ => panic!(),
    }
}

#[fixture]
fn check_config(
    #[default(None)] config: Option<PathBuf>,
    #[default(None)] lockfile: Option<Vec<PathBuf>>,
    #[default(None)] prefix: Option<Vec<PathBuf>>,
    #[default(None)] platform: Option<Vec<Platform>>,
    #[default(None)] environment: Option<Vec<String>>,
    #[default(None)] osi: Option<bool>,
    #[default(None)] ignore_pypi: Option<bool>,
    #[default(Some(OutputFormat::Default))] output: Option<OutputFormat>,
) -> CondaDenyCheckConfig {
    let cli = CondaDenyCliConfig::Check {
        lockfile,
        prefix,
        platform,
        environment,
        osi,
        ignore_pypi,
        output,
    };

    let config = get_config_options(config, cli);
    let config = config.unwrap();

    match config {
        CondaDenyConfig::Check(check_config) => check_config,
        _ => panic!(),
    }
}

#[fixture]
fn out() -> Vec<u8> {
    Vec::new()
}

#[rstest]
#[case("check", "test_default_use_case")]
#[case("list", "test_default_use_case")]
#[case("check", "test_default_use_case_pyproject")]
#[case("list", "test_default_use_case_pyproject")]
fn test_default_use_case(#[case] subcommand: &str, #[case] test_name: &str) {
    use core::str;

    let path_string = format!("tests/{}", test_name);
    let test_dir = Path::new(path_string.as_str());

    let output = Command::cargo_bin("conda-deny")
        .unwrap()
        .arg(subcommand)
        .current_dir(test_dir)
        .env("CLICOLOR_FORCE", "1")
        .output()
        .expect("Failed to execute command");

    let stdout = str::from_utf8(&output.stdout).unwrap();
    if subcommand == "check" {
        assert!(stdout.contains("There were \u{1b}[32m247\u{1b}[0m safe licenses and \u{1b}[31m301\u{1b}[0m unsafe licenses."), "{stdout}");
        output.assert().failure();
    } else {
        assert!(stdout.contains("\u{1b}[34mzstandard\u{1b}[0m \u{1b}[36m0.22.0\u{1b}[0m-\u{1b}[3;96mpy312h721a963_1\u{1b}[0m (\u{1b}[95mosx-arm64\u{1b}[0m): \u{1b}[33mBSD-3-Clause"));
        assert!(stdout.contains("\u{1b}[34mzlib\u{1b}[0m \u{1b}[36m1.3.1\u{1b}[0m-\u{1b}[3;96mh4ab18f5_1\u{1b}[0m (\u{1b}[95mlinux-64\u{1b}[0m): \u{1b}[33mZlib"));
        assert!(stdout.contains("\u{1b}[34mxz\u{1b}[0m \u{1b}[36m5.2.6\u{1b}[0m-\u{1b}[3;96mh166bdaf_0\u{1b}[0m (\u{1b}[95mlinux-64\u{1b}[0m): \u{1b}[33mLGPL-2.1 and GPL-2.0"));
        output.assert().success();
    }
}

#[test]
fn test_remote_whitelist_check() {
    // Create a temporary file for the license_whitelist.toml
    let mut temp_config_file = NamedTempFile::new().unwrap();
    let file_content = r#"[tool.conda-deny]
license-whitelist = "https://raw.githubusercontent.com/Quantco/conda-deny/refs/heads/main/tests/default_license_whitelist.toml""#;

    temp_config_file
        .as_file_mut()
        .write_all(file_content.as_bytes())
        .unwrap();

    let temp_config_file_path = temp_config_file.path().to_path_buf();

    let mut out = out();
    let check_config = check_config(
        Some(temp_config_file_path),
        Some(vec!["tests/default_pixi.lock".into()]),
        None,
        None,
        None,
        None,
        None,
        None,
    );

    let result = check(check_config, &mut out);
    let output = String::from_utf8(out).unwrap();

    assert!(output.contains(
        "There were \u{1b}[32m242\u{1b}[0m safe licenses and \u{1b}[31m299\u{1b}[0m unsafe licenses."
    ));
    assert!(result.is_err());
}

#[test]
fn test_multiple_whitelists_check() {
    // Create a temporary file for the license_whitelist.toml
    let mut temp_license_whitelist = NamedTempFile::new().unwrap();
    let file_content = r#"[tool.conda-deny]
                                safe-licenses = ["BSD-3-Clause"]"#;
    temp_license_whitelist
        .as_file_mut()
        .write_all(file_content.as_bytes())
        .unwrap();

    let temp_whitelist_path = temp_license_whitelist.path().to_path_buf();

    // Create a temporary file for pixi.toml
    let mut temp_pixi_toml = NamedTempFile::new().unwrap();
    let file_content = "[tool.conda-deny]
        license-whitelist = [
        \"https://raw.githubusercontent.com/Quantco/conda-deny/refs/heads/main/tests/default_license_whitelist.toml\",
        \"".to_string() + temp_whitelist_path.to_str().unwrap() + "\"]";

    temp_pixi_toml
        .as_file_mut()
        .write_all(file_content.as_bytes())
        .unwrap();

    let mut out = out();
    // Inject the temporary file's path into check_config
    let temp_path = Some(temp_pixi_toml.path().to_path_buf());
    let check_config = check_config(
        temp_path,
        Some(vec!["tests/default_pixi.lock".into()]),
        None,
        None,
        None,
        None,
        None,
        None,
    );

    let result = check(check_config, &mut out);
    let output = String::from_utf8(out).unwrap();

    assert!(output.contains(
        "There were \u{1b}[32m343\u{1b}[0m safe licenses and \u{1b}[31m198\u{1b}[0m unsafe licenses."
    ));
    assert!(result.is_err());
}

#[test]
fn test_platform_env_restrictions_check() {
    // Create a temporary file for pixi.toml
    let mut temp_pixi_toml = NamedTempFile::new().unwrap();
    let file_content = r#"[tool.conda-deny]
license-whitelist = "tests/default_license_whitelist.toml"
lockfile = "tests/default_pixi.lock"
platform = "linux-64"
environment = "lint""#;

    temp_pixi_toml
        .as_file_mut()
        .write_all(file_content.as_bytes())
        .unwrap();

    let mut out = out();
    // Inject the temporary file's path into check_config
    let temp_path = Some(temp_pixi_toml.path().to_path_buf());
    let check_config = check_config(
        temp_path,
        None,
        None,
        None,
        None,
        None,
        None,
        None,
    );

    let result = check(check_config, &mut out);
    let output = String::from_utf8(out).unwrap();

    assert!(output.contains(
        "There were \u{1b}[32m27\u{1b}[0m safe licenses and \u{1b}[31m21\u{1b}[0m unsafe licenses."
    ), "{output:?}");
    assert!(result.is_err());
}

#[test]
fn test_safe_licenses_in_config_check() {
    // Create a temporary file for pixi.toml
    let mut temp_pixi_toml = NamedTempFile::new().unwrap();
    let file_content = r#"[tool.conda-deny]
license-whitelist = "tests/default_license_whitelist.toml"
safe-licenses = ["BSD-3-Clause"]"#;

    temp_pixi_toml
        .as_file_mut()
        .write_all(file_content.as_bytes())
        .unwrap();

    let mut out = out();
    // Inject the temporary file's path into check_config
    let temp_path = Some(temp_pixi_toml.path().to_path_buf());
    let check_config = check_config(
        temp_path,
        Some(vec!["tests/default_pixi.lock".into()]),
        None,
        None,
        None,
        None,
        None,
        None,
    );

    let result = check(check_config, &mut out);
    let output = String::from_utf8(out).unwrap();

    assert!(output.contains(
        "There were \u{1b}[32m343\u{1b}[0m safe licenses and \u{1b}[31m198\u{1b}[0m unsafe licenses."
    ));
    assert!(result.is_err());
}

#[rstest]
fn test_osi_check(
    #[with(
        // CONFIG PATH
        None,
        // LOCKFILE PATHS
        Some(vec!["tests/default_pixi.lock".into()]),
        // PREFIXES
        None,
        // PLATFORM
        None,
        // ENVIRONMENT
        None,
        // OSI FLAG
        Some(true)
    )]
    check_config: CondaDenyCheckConfig,
    mut out: Vec<u8>,
) {
    let result = check(check_config, &mut out);

    let output = String::from_utf8(out).unwrap();

    assert!(output.contains(
        "There were \u{1b}[32m457\u{1b}[0m safe licenses and \u{1b}[31m84\u{1b}[0m unsafe licenses."
    ));
    assert!(result.is_err());
}

#[rstest]
fn test_prefix_list(
    #[with(
        // CONFIG PATH
        None,
        // LOCKFILE PATHS
        None,
        // PREFIXES
        Some(vec!["tests/test_conda_prefixes/test-env".into()])
)]
    list_config: CondaDenyListConfig,
    mut out: Vec<u8>,
) {
    // When --prefix is specified, only the license information for the conda-meta directory in the specified prefix should be listed
    // License information from pixi.lock should not be listed
    let result = list(list_config, &mut out);
    assert!(result.is_ok(), "{:?}", result.unwrap_err());
    let output = String::from_utf8(out).unwrap();
    let line_count = output.split("\n").count();
    let expected_line_count = 51;
    assert_eq!(
        line_count, expected_line_count,
        "Unexpected number of output lines"
    );
}

#[rstest]
fn test_exception_check(
    #[with(Some(PathBuf::from(
        // CONFIG PATH
        "tests/test_exception_use_case/config_with_exception.toml")),
        // LOCKFILE PATHS
        Some(vec!["tests/default_pixi.lock".into()])
    )]
    check_config: CondaDenyCheckConfig,
    mut out: Vec<u8>,
) {
    let result = check(check_config, &mut out);

    let output = String::from_utf8(out).unwrap();

    assert!(output.contains(
        "There were \u{1b}[32m527\u{1b}[0m safe licenses and \u{1b}[31m14\u{1b}[0m unsafe licenses."
    ), "{output:?}");
    assert!(result.is_err());
}

#[rstest]
fn test_pypi_ignore_check(
    #[with(
        // CONFIG PATH
        Some(PathBuf::from("tests/test_pypi_ignore/pixi.toml")),
        // LOCKFILE PATHS
        Some(vec!["tests/test_pypi_ignore/lockfile_with_pypi_packages.lock".into()]),
        // PREFIXES
        None,
        // PLATFORM
        None,
        // ENVIRONMENT
        None,
        // OSI FLAG
        None,
        // IGNORE PYPI
        Some(true)
    )]
    check_config: CondaDenyCheckConfig,
    mut out: Vec<u8>,
) {
    let result = check(check_config, &mut out);

    let output = String::from_utf8(out).unwrap();

    assert!(output.contains(
        "There were \u{1b}[32m5\u{1b}[0m safe licenses and \u{1b}[31m17\u{1b}[0m unsafe licenses."
    ));
    assert!(result.is_err());
}

#[rstest]
fn test_pypi_ignore_error(
    #[with(
        // CONFIG PATH
        Some(PathBuf::from("tests/test_pypi_ignore/pixi.toml")),
        // LOCKFILE PATHS
        Some(vec!["tests/test_pypi_ignore/lockfile_with_pypi_packages.lock".into()]),
        // PREFIXES
        None,
        // PLATFORM
        None,
        // ENVIRONMENT
        None,
        // OSI FLAG
        None,
        // IGNORE PYPI
        Some(false)
    )]
    check_config: CondaDenyCheckConfig,
    mut out: Vec<u8>,
) {
    let result = check(check_config, &mut out);
    if let Err(e) = &result {
        println!("Actual error: {}", e);
    }
    assert!(result.is_err());
    assert!(format!("{:?}", result).contains("Pypi packages are not supported: beautifulsoup4"));
    assert_eq!(out, b"");
}

#[rstest]
fn test_bundle_prefix() {
    let mut out = out();
    let temp_dir = tempfile::tempdir().unwrap();
    let bundle_config = bundle_config(
        // CONFIG PATH
        None,
        // LOCKFILE PATHS
        None,
        // PREFIXES
        Some(vec!["tests/test_conda_prefixes/test-env".into()]),
        // PLATFORM
        None,
        // ENVIRONMENT
        None,
        // IGNORE PYPI
        None,
        // DIRECTORY
        Some(temp_dir.path().join(Path::new("test_bundle"))),
    );

    // Suppress progress bar
    std::env::set_var("NO_PROGRESS", "1");

    bundle(bundle_config.clone(), &mut out).unwrap();
    let bundle_dir = bundle_config.directory.unwrap();

    let mut files = Vec::new();
    let mut dirs = Vec::new();

    for e in WalkDir::new(&bundle_dir).into_iter().filter_map(|e| e.ok()) {
        if e.metadata().unwrap().is_file() {
            files.push(e.path().to_path_buf());
        }
        if e.metadata().unwrap().is_dir() {
            dirs.push(e.path().to_path_buf());
        }
    }
    assert_eq!(files.len(), 48, "{files:?}");
    assert_eq!(dirs.len(), 52, "{dirs:?}");

    let top_level_entries = std::fs::read_dir(bundle_dir).unwrap();
    let top_level_entry_count = top_level_entries.count();
    assert_eq!(top_level_entry_count, 47);
    let output = String::from_utf8(out).unwrap();
    assert!(output.contains("License files written to:"));
}

#[rstest]
fn test_bundle_lockfile() {
    let mut out = out();
    let temp_dir = tempfile::tempdir().unwrap();
    let bundle_config = bundle_config(
        // CONFIG PATH
        None,
        // LOCKFILE PATHS
        Some(vec!["tests/test_default_use_case/pixi.lock".into()]),
        // PREFIXES
        None,
        // PLATFORM
        None,
        // ENVIRONMENT
        None,
        // IGNORE PYPI
        None,
        // DIRECTORY
        Some(temp_dir.path().join(Path::new("test_bundle"))),
    );

    // Suppress progress bar
    std::env::set_var("NO_PROGRESS", "1");

    bundle(bundle_config.clone(), &mut out).unwrap();
    let bundle_dir = bundle_config.directory.unwrap();

    let mut files = Vec::new();
    let mut dirs = Vec::new();

    for e in WalkDir::new(&bundle_dir).into_iter().filter_map(|e| e.ok()) {
        if e.metadata().unwrap().is_file() {
            files.push(e.path().to_path_buf());
        }
        if e.metadata().unwrap().is_dir() {
            dirs.push(e.path().to_path_buf());
        }
    }
    assert_eq!(files.len(), 6649);
    assert_eq!(dirs.len(), 4724);

    let top_level_entries = std::fs::read_dir(bundle_dir).unwrap();
    let top_level_entry_count = top_level_entries.count();
    assert_eq!(top_level_entry_count, 356);
    let output = String::from_utf8(out).unwrap();
    assert!(output.contains("License files written to:"));
}
