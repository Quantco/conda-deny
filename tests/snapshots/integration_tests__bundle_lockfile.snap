---
source: tests/integration_tests.rs
expression: pretty_output
---
bzip2-1.0.8-h99b78c6_7
bzip2-1.0.8-h99b78c6_7/LICENSE
ca-certificates-2025.8.3-hbd8a1cb_0
ca-certificates-2025.8.3-hbd8a1cb_0/LICENSE
delve-1.25.1-hce30654_0
delve-1.25.1-hce30654_0/LICENSE
delve-1.25.1-hce30654_0/license-files
delve-1.25.1-hce30654_0/license-files/github.com
delve-1.25.1-hce30654_0/license-files/github.com/cosiner
delve-1.25.1-hce30654_0/license-files/github.com/cosiner/argv
delve-1.25.1-hce30654_0/license-files/github.com/cosiner/argv/LICENSE
delve-1.25.1-hce30654_0/license-files/github.com/cpuguy83
delve-1.25.1-hce30654_0/license-files/github.com/cpuguy83/go-md2man
delve-1.25.1-hce30654_0/license-files/github.com/cpuguy83/go-md2man/v2
delve-1.25.1-hce30654_0/license-files/github.com/cpuguy83/go-md2man/v2/md2man
delve-1.25.1-hce30654_0/license-files/github.com/cpuguy83/go-md2man/v2/md2man/LICENSE.md
delve-1.25.1-hce30654_0/license-files/github.com/derekparker
delve-1.25.1-hce30654_0/license-files/github.com/derekparker/trie
delve-1.25.1-hce30654_0/license-files/github.com/derekparker/trie/LICENSE
delve-1.25.1-hce30654_0/license-files/github.com/go-delve
delve-1.25.1-hce30654_0/license-files/github.com/go-delve/delve
delve-1.25.1-hce30654_0/license-files/github.com/go-delve/delve/LICENSE
delve-1.25.1-hce30654_0/license-files/github.com/go-delve/liner
delve-1.25.1-hce30654_0/license-files/github.com/go-delve/liner/COPYING
delve-1.25.1-hce30654_0/license-files/github.com/google
delve-1.25.1-hce30654_0/license-files/github.com/google/go-dap
delve-1.25.1-hce30654_0/license-files/github.com/google/go-dap/LICENSE
delve-1.25.1-hce30654_0/license-files/github.com/hashicorp
delve-1.25.1-hce30654_0/license-files/github.com/hashicorp/golang-lru
delve-1.25.1-hce30654_0/license-files/github.com/hashicorp/golang-lru/simplelru
delve-1.25.1-hce30654_0/license-files/github.com/hashicorp/golang-lru/simplelru/LICENSE
delve-1.25.1-hce30654_0/license-files/github.com/hashicorp/golang-lru/simplelru/simplelru
delve-1.25.1-hce30654_0/license-files/github.com/hashicorp/golang-lru/simplelru/simplelru/lru.go
delve-1.25.1-hce30654_0/license-files/github.com/hashicorp/golang-lru/simplelru/simplelru/lru_interface.go
delve-1.25.1-hce30654_0/license-files/github.com/mattn
delve-1.25.1-hce30654_0/license-files/github.com/mattn/go-isatty
delve-1.25.1-hce30654_0/license-files/github.com/mattn/go-isatty/LICENSE
delve-1.25.1-hce30654_0/license-files/github.com/mattn/go-runewidth
delve-1.25.1-hce30654_0/license-files/github.com/mattn/go-runewidth/LICENSE
delve-1.25.1-hce30654_0/license-files/github.com/rivo
delve-1.25.1-hce30654_0/license-files/github.com/rivo/uniseg
delve-1.25.1-hce30654_0/license-files/github.com/rivo/uniseg/LICENSE.txt
delve-1.25.1-hce30654_0/license-files/github.com/russross
delve-1.25.1-hce30654_0/license-files/github.com/russross/blackfriday
delve-1.25.1-hce30654_0/license-files/github.com/russross/blackfriday/v2
delve-1.25.1-hce30654_0/license-files/github.com/russross/blackfriday/v2/LICENSE.txt
delve-1.25.1-hce30654_0/license-files/github.com/spf13
delve-1.25.1-hce30654_0/license-files/github.com/spf13/cobra
delve-1.25.1-hce30654_0/license-files/github.com/spf13/cobra/LICENSE.txt
delve-1.25.1-hce30654_0/license-files/github.com/spf13/pflag
delve-1.25.1-hce30654_0/license-files/github.com/spf13/pflag/LICENSE
delve-1.25.1-hce30654_0/license-files/go.starlark.net
delve-1.25.1-hce30654_0/license-files/go.starlark.net/LICENSE
delve-1.25.1-hce30654_0/license-files/golang.org
delve-1.25.1-hce30654_0/license-files/golang.org/x
delve-1.25.1-hce30654_0/license-files/golang.org/x/arch
delve-1.25.1-hce30654_0/license-files/golang.org/x/arch/LICENSE
delve-1.25.1-hce30654_0/license-files/golang.org/x/sync
delve-1.25.1-hce30654_0/license-files/golang.org/x/sync/errgroup
delve-1.25.1-hce30654_0/license-files/golang.org/x/sync/errgroup/LICENSE
delve-1.25.1-hce30654_0/license-files/golang.org/x/sys
delve-1.25.1-hce30654_0/license-files/golang.org/x/sys/unix
delve-1.25.1-hce30654_0/license-files/golang.org/x/sys/unix/LICENSE
delve-1.25.1-hce30654_0/license-files/golang.org/x/telemetry
delve-1.25.1-hce30654_0/license-files/golang.org/x/telemetry/LICENSE
delve-1.25.1-hce30654_0/license-files/gopkg.in
delve-1.25.1-hce30654_0/license-files/gopkg.in/yaml.v3
delve-1.25.1-hce30654_0/license-files/gopkg.in/yaml.v3/LICENSE
delve-1.25.1-hce30654_0/license-files/gopkg.in/yaml.v3/NOTICE
icu-75.1-hfee45f7_0
icu-75.1-hfee45f7_0/LICENSE
libblas-3.9.0-34_h10e41b3_openblas
libblas-3.9.0-34_h10e41b3_openblas/LICENSE.txt
libcblas-3.9.0-34_hb3479ef_openblas
libcblas-3.9.0-34_hb3479ef_openblas/LICENSE.txt
libcxx-20.1.8-hf598326_1
libcxx-20.1.8-hf598326_1/libcxx
libcxx-20.1.8-hf598326_1/libcxx/LICENSE.TXT
libexpat-2.7.1-hec049ff_0
libexpat-2.7.1-hec049ff_0/COPYING
libffi-3.4.6-h1da3d7d_1
libffi-3.4.6-h1da3d7d_1/LICENSE
libgfortran-15.1.0-hfdf1602_0
libgfortran-15.1.0-hfdf1602_0/COPYING3
libgfortran5-15.1.0-hb74de2c_0
libgfortran5-15.1.0-hb74de2c_0/COPYING3
liblapack-3.9.0-34_hc9a63f6_openblas
liblapack-3.9.0-34_hc9a63f6_openblas/LICENSE.txt
liblzma-5.8.1-h39f12f2_2
liblzma-5.8.1-h39f12f2_2/COPYING
liblzma-5.8.1-h39f12f2_2/COPYING.0BSD
libmpdec-4.0.0-h5505292_0
libmpdec-4.0.0-h5505292_0/COPYRIGHT.txt
libopenblas-0.3.30-openmp_h60d53f8_2
libopenblas-0.3.30-openmp_h60d53f8_2/LICENSE
libopenblas-0.3.30-openmp_h60d53f8_2/lapack-netlib
libopenblas-0.3.30-openmp_h60d53f8_2/lapack-netlib/LICENSE
libzlib-1.3.1-h8359307_2
libzlib-1.3.1-h8359307_2/LICENSE
llvm-openmp-20.1.8-hbb9b287_2
llvm-openmp-20.1.8-hbb9b287_2/openmp
llvm-openmp-20.1.8-hbb9b287_2/openmp/LICENSE.TXT
ncurses-6.5-h5e97a16_3
ncurses-6.5-h5e97a16_3/COPYING
numpy-2.3.2-py313haac90e2_1
numpy-2.3.2-py313haac90e2_1/LICENSE.txt
openssl-3.5.2-he92f556_0
openssl-3.5.2-he92f556_0/LICENSE.txt
python-3.13.5-hf3f3da0_102_cp313
python-3.13.5-hf3f3da0_102_cp313/LICENSE
python_abi-3.13-8_cp313
python_abi-3.13-8_cp313/LICENSE
readline-8.2-h1d1bf99_2
readline-8.2-h1d1bf99_2/COPYING
tk-8.6.13-h892fb3f_2
tk-8.6.13-h892fb3f_2/tcl8.6.13
tk-8.6.13-h892fb3f_2/tcl8.6.13/license.terms
tzdata-2025b-h78e105d_0
tzdata-2025b-h78e105d_0/LICENSE
